### PR TITLE
Make ZST checks in core/alloc more readable

### PR DESCRIPTION
**Pull request by scottmcm**
There's a bunch of these checks because of special handing for ZSTs in various unsafe implementations of stuff.

This lets them be `T::IS_ZST` instead of `mem::size_of::<T>() == 0` every time, making them both more readable and more terse.

*Not* proposed for stabilization.  Would be `pub(crate)` except `alloc` wants to use it too.

(And while it doesn't matter now, if we ever get something like #85836 making it a const can help codegen be simpler.)

### CODE REVIEW

**Comment by thomcc** [comment]
Hm, neat. I know you said it shouldn't impact codegen at all, but I'm a little curious.

@bors try @rust-timer queue

**Comment by scottmcm** [comment]
> I know you said it shouldn't impact codegen at all

All I meant by that was that there's currently no special handling for `switchInt` on a constant at the LLVM level.

I agree this should have a perf run, but was waiting on the PR build.  Looks like when I fixed some of the warnings(-as-errors) about unused imports I broke the docs 🤦

**Comment by scottmcm** [comment]
Hmm, that's a weird build failure.  RustDoc handled it fine, but now it's a linkchecker error?

Local repro time, I guess -- `x.py doc library/std` was clean :(

**Comment by scottmcm** [comment]
Ok, filed #102221 for the weird failure when I tried to use `cfg(doc)`, and fixed it with just using an `#[allow]`.

Should be good to go now 🤞

@rustbot ready

File: library/core/src/mem/mod.rs [hunk 1]
```diff
@@ -1178,3 +1178,18 @@ pub const fn discriminant<T>(v: &T) -> Discriminant<T> {
 pub const fn variant_count<T>() -> usize {
     intrinsics::variant_count::<T>()
 }
+
+/// This is here only to simplify all the ZST checks we need in the library.
+/// It's not on a stabilization track right now.
```
Comment (thomcc) [inline library/core/src/mem/mod.rs:1183]: Can you add some more documentation for using this? Maybe a small example? E.g. I'd like to be able to just say "this should use `core::mem::SizedTypeProperties::IS_ZST`" in a review, without having to explain much more.

**Comment by thomcc** [comment]
Thank you.

@bors r+

**Pull request merged**